### PR TITLE
Merge confgen & main models (#254)

### LINKED ISSUE #254: Reuse API data structures in confgen
The config created by `confgen` is generated from nested `map[string]interface{}` rather than reusing the structs defined in [api](https://github.com/netobserv/flowlogs-pipeline/tree/main/pkg/api) and [config](https://github.com/netobserv/flowlogs-pipeline/blob/main/pkg/config/config.go)
https://github.com/netobserv/flowlogs-pipeline/blob/ba915706711650425f17c9402ffda2b9d8841ed7/pkg/confgen/flowlogs2metrics_config.go#L28-L176

The pipeline config builder could also become handy
https://github.com/netobserv/flowlogs-pipeline/blob/main/pkg/config/pipeline_builder.go

### PR DESCRIPTION
**Pull request by jotak**
Use main model in confgen, use pipeline builder

Note, changes required in NOO

Fixes #254

- Confgen: init from options (avoid many globals, so it's less error prone on consumers)
- Removed duplicated confgen model
- Use PipelineBuilder to simplify pipeline generation
- Make confgen easier to consume as a lib (e.g. for NOO)
  - Do not make it necessary to call "Run": parsing definition files
    should be sufficient
  - Do not make it necessary to work with files written on disk: work
    with []byte instead
- SkipWithTags should not result in an error when a file is skipped
- Add confgen tests

### CODE REVIEW

File: pkg/confgen/flowlogs2metrics_config.go [hunk 1]
```diff
@@ -22,163 +22,70 @@ import (
...
+		"pipeline":   pipeline.GetStages(),
+		"parameters": pipeline.GetStageParams(),
 	}
-	return configStruct
 }
 
-func (cg *ConfGen) GenerateTruncatedConfig(stages []string) config.ConfigFileStruct {
-	parameters := make([]config.StageParam, len(stages))
-	for i, stage := range stages {
+func (cg *ConfGen) GenerateTruncatedConfig() []config.StageParam {
```
Comment (ronensc) [inline pkg/confgen/flowlogs2metrics_config.go:63]: This signature change requires changes in NOO.

**Review by ronensc — APPROVED** [review]
Thanks for refactoring and cleaning up the code :)

File: pkg/confgen/flowlogs2metrics_config.go [hunk 2]
```diff
@@ -22,163 +22,70 @@ import (
...
+			Metrics: cg.promMetrics,
+		})
+	}
+	if cg.config.Write.Loki != nil {
+		next.WriteLoki("write_loki", *cg.config.Write.Loki)
+	}
+	return map[string]interface{}{
+		"log-level":  "error",
+		"pipeline":   pipeline.GetStages(),
+		"parameters": pipeline.GetStageParams(),
```
[the remark this answers is not in the document]
Comment (jotak) [reply pkg/confgen/flowlogs2metrics_config.go:59]: Good point I'll take it back.

[the remark this answers is not in the document]
Comment (jotak) [reply pkg/confgen/flowlogs2metrics_config.go:59]: (but now I want to also remove globals from `config` 😅

File: pkg/confgen/flowlogs2metrics_config.go (hunk 1 above)
Comment (jotak) [reply pkg/confgen/flowlogs2metrics_config.go:63]: Yes I have this PR on NOO https://github.com/netobserv/network-observability-operator/pull/142

File: pkg/confgen/flowlogs2metrics_config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jotak) [reply pkg/confgen/flowlogs2metrics_config.go:59]: done

File: pkg/confgen/confgen_test.go [hunk 3]
```diff
@@ -116,34 +143,110 @@ visualization:
...
+func Test_RunShortConfGen(t *testing.T) {
+	// Prepare
+	dirPath := "/tmp/getDefinitionFilesTest"
+	cg := NewConfGen(&Options{
+		SrcFolder:                dirPath,
+		DestConfFile:             "/tmp/destConfigTest",
+		DestDocFile:              "/tmp/destDocTest",
+		DestGrafanaJsonnetFolder: "/tmp/destJsonnetTest",
+	})
+	err := os.MkdirAll(dirPath, 0755)
```
[the remark this answers is not in the document]
Comment (jotak) [reply pkg/confgen/confgen_test.go]: done (deferred cleanup)

File: pkg/config/pipeline_builder.go [hunk 4]
```diff
@@ -80,22 +96,22 @@ func (b *PipelineBuilderStage) next(name string, param StageParam) PipelineBuild
...
+	return b.next(name, NewTransformFilterParams(name, filter))
 }
 
 // TransformNetwork chains the current stage with a TransformNetwork stage and returns that new stage
 func (b *PipelineBuilderStage) TransformNetwork(name string, nw api.TransformNetwork) PipelineBuilderStage {
-	return b.next(name, StageParam{Name: name, Transform: &Transform{Type: api.NetworkType, Network: &nw}})
+	return b.next(name, NewTransformNetworkParams(name, nw))
 }
 
 // ConnTrack chains the current stage with a ConnTrack stage and returns that new stage
```
[the remark this answers is not in the document]
Comment (jotak) [reply pkg/config/pipeline_builder.go:117]: done

**Comment by jotak** [comment]
@ronensc feedback addressed .. I actually ended up doing more changes in order to remove globals from `config.go`. They are replaced with explicit arguments, which helps to understand what's expected from the consuming side

File: pkg/confgen/flowlogs2metrics_config.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ronensc) [reply pkg/confgen/flowlogs2metrics_config.go]: I think the replacement of the `Output` type in the tests was missed.

File: pkg/config/config.go [hunk 5]
```diff
@@ -101,20 +95,22 @@ type Write struct {
 }
 
 // ParseConfig creates the internal unmarshalled representation from the Pipeline and Parameters json
-func ParseConfig() error {
-	logrus.Debugf("config.Opt.PipeLine = %v ", Opt.PipeLine)
-	err := json.Unmarshal([]byte(Opt.PipeLine), &PipeLine)
+func ParseConfig(opts Options) (ConfigFileStruct, error) {
+	out := ConfigFileStruct{}
+
+	logrus.Debugf("config.Opt.PipeLine = %v ", opts.PipeLine)
```
Comment (ronensc) [inline pkg/config/config.go:101]:
```suggestion
	logrus.Debugf("opts.PipeLine = %v ", opts.PipeLine)
```

**Pull request merged**